respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,140 @@
+
+    Parameters
+    ----------
+    params : pandas.DataFrame
+        DataFrame containing model parameters.
+    options : dict
+        Dictionary containing model options.
+    moments : np.ndarray
+        Array with shape (n_moments,) containing the moments of the actual data.
+    calc_moments : callable
+        Function to compute the moments from the simulated data.
+    weighting_matrix : np.ndarray
+        Array with shape (n_moments, n_moments) which is used to weight the squared
+        errors of moments.
+    replacement : callable or int
+        If int then all missing moments of the evaluation with the current moments
+         are set to this int. Else callable that maps missing indices into
+          replacement int.
+    all_dims: boolean
+        If set to false objective is returned. If set to True weighted array
+        of component deviations will be returned.
+
+    Returns
+    -------
+    msm_function : :func:`msm`
+        Criterion function where all arguments except the parameter vector are set.
+
+    """
+    optim_paras, options = process_params_and_options(params, options)
+
+    simulate = get_simulate_func(params, options)
+
+    if weighting_matrix is None:
+        weighting_matrix = pd.DataFrame(
+            np.eye(len(moments)), index=moments.index, columns=moments.index
+        )
+
+    msm_function = functools.partial(
+        msm,
+        simulate=simulate,
+        moments=moments,
+        weighting_matrix=weighting_matrix,
+        calc_moments=calc_moments,
+        replacement=replacement,
+        all_dims=all_dims,
+    )
+
+    return msm_function
+
+
+def msm(
+    params, simulate, moments, weighting_matrix, calc_moments, replacement, all_dims
+):
+    """Criterion function for the estimation with MSM.
+
+    This function calculates the sum of weighted squared errors of moments.
+
+    Parameters
+    ----------
+    params : pandas.Series
+        Parameter Series.
+    simulate : :func:`~respy.simulate.simulate`
+        Function to simulate a new data set for a given set of parameters.
+    moments : np.ndarray
+        Array with shape (n_moments,) containing the moments from the actual data.
+    weighting_matrix : np.ndarray
+        Array with shape (n_moments, n_moments) which is used to weight the squared
+        errors of moments.
+    calc_moments : callable
+        Function to compute the moments from the simulated data.
+    replacement : callable or int
+        If int then all missing moments of the evaluation with the current
+        moments are set to this int. Else callable that maps missing
+        indices into replacement int.
+    all_dims: boolean
+        If set to false objective is returned.
+        If set to True weighted array of component deviations will be returned.
+    """
+    df = simulate(params)
+
+    estimated_moments = calc_moments(df)
+
+    # Get index
+    idx = moments.index.intersection(estimated_moments.index)
```

Comment:
Correct me if I am wrong, but if some moments do no exist in `estimated_moments`, you select a subset of the empirical moments.

This step should actually be a function which receives the empirical and simulated moments. Then, the function distinguishes between the cases when moments are single objects, lists or dictionaries. Then, use index and columns from the empirical moments and apply them to the simulated moments with `simulated_moments.reindex(index=index, columns=columns)`. It automatically creates NaNs for missing indices or columns.

Then, maybe another function which flattens the moments. If the object is a dataframe, join the index and the column name via `_` and turn it into a series. Concat all series.